### ISSUE TITLE
Getting return values of the scheduled function

### ISSUE THREAD

**Issue by salmanshahid8** [dbader/schedule#327]
Hi all,
So I have been using **schedule** for a couple of months. However since a few days I am looking for a solution to get the returned values of the function that executes at a scheduled time. Here is a sample function.

```
def get_data(dir):
    df = pd.DataFrame()
    file_name = "nofile"
    # The rest of the body of the code executes and updates the two variables df and file_name
    return df, file_name
```

What I want to do is to schedule this function at 6:00 AM. If the _file_name_ remains "nofile" after the execution of "get_data()" at 6:00 AM, the function should be re-executed at 6:00 PM. So like this

```
 def sched():
    schedule.every().day.at().hour(6).do(get_data(), [dir])
    # Here I cannot get the file_name as return value of get_data()
    if file_name == "nofile":
        schedule.every().day.at().hour(18).do(get_data(), [dir])
        file_name = "nofile"
```

I have no idea how to do it with schedule. Please help.

**Comment by mralecthomas** [comment]
I would schedule two jobs to handle this:

`schedule.every().day.at("06:00").do(get_data(), [dir])`
`schedule.every().day.at("18:00").do(get_data(), [dir])`

**Comment by V-for-Vaggelis** [comment]
Did you find a solution for this? I'm simply trying to return some data from the job function but I can't access them after the job is executed.

**Comment by knowsuchagency** [comment]
Use a global queue or cache i.e.

```python
queue = []

def f1():
  """scheduled first"""
  queue.append(result)

def f2():
  """scheduled later"""
  if queue:
    result = queue.pop()
  ...
```

**Comment by SijmenHuizenga** [comment]
Hi! Thanks for the great question. 

Firstly, keep in mind that example code in the question does not work:
```python
schedule.every().day.at().hour(6).do(get_data(), [dir])
```
This will throw an `Invalid unit` error because no value is provided in the `.at()` statement. To schedule a job to run at 6 AM, use:
```python
schedule.every().day.at("06:00").do(job)
```

Getting return values from a job is not supported by the scheduler. The concept of a job is that it is a self-contained piece of code.  If it needs to store data to be processed by some other piece of code you can use a variable outside the job's scope. The example provided by @knowsuchagency shows how this can be achieved. We should probably add such an example to the docs.

Now, let's take a look at your use-case:

> What I want to do is to schedule this function at 6:00 AM. If the file_name remains "nofile" after the execution of "get_data()" at 6:00 AM, the function should be re-executed at 6:00 PM.

One way to implement this without global variables would be:

```python
def get_data():
    file_name = "nofile"
    # The rest of the body of the code executes and updates the two file_name
    return file_name

def job_morning():
    file_name = get_data()

    if file_name == "nofile":
        # If the file_name remains "nofile" after 
        # the execution of "get_data()" at 6:00 AM,
        # the function should be re-executed at 6:00 PM.
        schedule.every().day.at("18:00").do(job_evening)

def job_evening():
    get_data()
    
    # Cancel the evening job as it will be re-scheduled by
    # the morning job if it is needs to run again tomorrow.
    return schedule.CancelJob

# Run the morning job every day at 6 AM
schedule.every().day.at("06:00").do(job_evening)
```

**Comment by SijmenHuizenga** [comment]
In general; I don't think supporting storage and retrieval of return-values is something we should do.

There are many different ways people may want to use return values. Some only need the last return value, some need all values in a queue-like manner and some may only sometimes want to sometimes store return values. If we store multiple return values, the next question in line will be: when was this return-value created? So then we would have to store metadata about which executions return what values. And when are the stored return values removed? How can we protect users from introducing memory leaks? All these answers can be answered but it is quite an increase of complexity.

On the other hand, storing return-values yourself is quite easy: instead of `return x`, do `global x; x = 3` or for class methods `self.x = 3`. You have all the freedom to build it however you like. There is not a huge gain when schedule is storing these return values for you.

So basically, I think the added value of return values is too low for the complexity it introduces.

If you do have a more advanced use-case where storing return values in some advanced manner is required, it might be good to investigate moving to some more advanced scheduling system. Schedule is designed to be a simple solution for your relatively simple scheduling needs. If you need highly complex scheduling a different library will probably help you more.

I hope this explains why I think Schedule should not support return values. If you have more questions or have a different opinion feel free to continue the conversation. Re-open is always a possibility :)
